File: solidarity/Services/ZK/ZKLogger.swift
//
//  ZKLogger.swift
//  solidarity
//
//  Lightweight unified logging for Semaphore-related flows.
//

import Foundation
import os

enum ZKLog {
  private static let logger = Logger(subsystem: AppBranding.currentLoggerSubsystem, category: "Semaphore")

  /// Caller-supplied messages may contain commitment prefixes, scopes, or
  /// other identifiers that should not leak verbatim from a release build.
  /// We mark the payload as `.private` for the unified logger (so it is
  /// redacted in non-debug log streams) and additionally drop the stdout
  /// `print(...)` mirror unless we are in a DEBUG build.
  static func info(_ message: String) {
    logger.info("\(message, privacy: .private)")
    #if DEBUG
    print("[Semaphore] \(message)")
    #endif
  }

  static func error(_ message: String) {
    logger.error("\(message, privacy: .private)")
    #if DEBUG
    print("[Semaphore][Error] \(message)")
    #endif
  }

  /// Operational status messages with no identifier content. Safe to leave
  /// public so we keep release-build observability for things like
  /// "Semaphore prover unavailable, falling back".
  static func status(_ message: String) {
    logger.info("\(message, privacy: .public)")
    #if DEBUG
    print("[Semaphore] \(message)")
    #endif
  }
}

File: apps/expo/src/zk/zkLogger.ts
/**
 * ZK logger — TS port of solidarity/Services/ZK/ZKLogger.swift.
 *
 * Mirrors the Swift `ZKLog` static API (`info`, `error`, `status`). On
 * device the Swift version is backed by `os.Logger` with `.private`
 * redaction. We don't have that primitive on RN, so the equivalent is:
 *   - DEV  : `console.log` mirror with the same `[Semaphore]` prefix.
 *   - PROD : drop `info` (caller content may contain commitment prefixes
 *            / scopes) and route warn+error to `eventRepository.record`
 *            so they show up in the audit log without leaking via stdout.
 *
 * `status` is operational text with no identifier content, so it's safe
 * to keep visible in release builds.
 */
import { record as recordEvent } from '@/feedback/eventRepository';

export type ZkLogLevel = 'debug' | 'info' | 'warn' | 'error';

const PREFIX = '[Semaphore]';

function isDev(): boolean {
  return typeof __DEV__ !== 'undefined' ? __DEV__ : process.env.NODE_ENV !== 'production';
}

export const zkLogger = {
  /** Verbose tracing. DEV only. */
  debug(message: string, data?: unknown): void {
    if (isDev()) {
      console.warn(`${PREFIX}[Debug] ${message}`, data ?? '');
    }
  },

  /**
   * Caller-supplied messages may contain commitment prefixes / scopes /
   * identifiers — kept out of release stdout. Routed to eventRepository
   * at debug level only (i.e. dropped) to match Swift `.private` redaction.
   */
  info(message: string, data?: unknown): void {
    if (isDev()) {
      console.warn(`${PREFIX} ${message}`, data ?? '');
    }
  },

  /** Warnings — release builds append to the audit log. */
  warn(message: string, data?: unknown): void {
    if (isDev()) {
      console.warn(`${PREFIX}[Warn] ${message}`, data ?? '');
    } else {
      recordEvent('zk', { level: 'warn', message, data });
    }
  },

  /** Errors — release builds append to the audit log. */
  error(message: string, data?: unknown): void {
    if (isDev()) {
      console.error(`${PREFIX}[Error] ${message}`, data ?? '');
    } else {
      recordEvent('zk', { level: 'error', message, data });
    }
  },

  /**
   * Operational status messages with no identifier content. Safe to leave
   * visible across build flavours (mirrors Swift `status(_:)`).
   */
  status(message: string): void {
    console.warn(`${PREFIX} ${message}`);
  },

  /** Unified entry point used by the spec contract: `log(level, message, data?)`. */
  log(level: ZkLogLevel, message: string, data?: unknown): void {
    switch (level) {
      case 'debug':
        zkLogger.debug(message, data);
        return;
      case 'info':
        zkLogger.info(message, data);
        return;
      case 'warn':
        zkLogger.warn(message, data);
        return;
      case 'error':
        zkLogger.error(message, data);
        return;
    }
  },
};

export function log(level: ZkLogLevel, message: string, data?: unknown): void {
  zkLogger.log(level, message, data);
}
